lib/db/schema/patch-045-046.sql

```diff
@@ -0,0 +1,188 @@
+  verifiedAt BIGINT UNSIGNED,
+  createdAt BIGINT UNSIGNED NOT NULL
+) ENGINE=InnoDB;
+
+DROP procedure IF EXISTS `createEmail_1`;
+
+CREATE PROCEDURE `createEmail_1` (
+    IN `normalizedEmail` VARCHAR(255),
+    IN `email` VARCHAR(255),
+    IN `uid` BINARY(16) ,
+    IN `emailCode` BINARY(16),
+    IN `isVerified` TINYINT(1),
+    IN `isPrimary` TINYINT(1),
+    IN `verifiedAt` BIGINT UNSIGNED,
+    IN `createdAt` BIGINT UNSIGNED
+)
+BEGIN
+    DECLARE EXIT HANDLER FOR SQLEXCEPTION
+        BEGIN
+        ROLLBACK;
+        RESIGNAL;
+    END;
+
+    START TRANSACTION;
+
+    -- Currently, can not add an email that is specified in the
+    -- accounts table, regardless of verification state.
+    SET @emailExists = 0;
+    SELECT COUNT(*) INTO @emailExists FROM accounts a WHERE a.normalizedEmail = normalizedEmail;
+    IF @emailExists > 0 THEN
+        SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Email already exists in accounts table.';
+    END IF;
+
+    INSERT INTO emails(
+        normalizedEmail,
+        email,
+        uid,
+        emailCode,
+        isVerified,
+        isPrimary,
+        verifiedAt,
+        createdAt
+    )
+    VALUES(
+        LOWER(normalizedEmail),
+        email,
+        uid,
+        emailCode,
+        isVerified,
+        isPrimary,
+        verifiedAt,
+        createdAt
+    );
+
+    COMMIT;
+END;
+
+CREATE PROCEDURE `accountEmails_1` (
+    IN `inUid` BINARY(16)
+)
+BEGIN
+    -- Return all emails that belong to the uid.
+    -- Email from the accounts table is considered the primary email.
+	(SELECT
+        a.normalizedEmail,
+        a.email,
+        a.uid,
+        a.emailCode,
+        a.emailVerified AS isVerified,
+        TRUE AS isPrimary,
+        a.createdAt AS verifiedAt,
+        a.createdAt
+    FROM
+        accounts a
+    WHERE
+        uid = LOWER(inUid))
+
+    UNION ALL
+
+    (SELECT
+        e.normalizedEmail,
+        e.email,
+        e.uid,
+        e.emailCode,
+        e.isVerified,
+        e.isPrimary,
+        e.verifiedAt,
+        e.createdAt
+    FROM
+        emails e
+    WHERE
+        uid = LOWER(inUid));
+END;
+
+CREATE PROCEDURE `deleteEmail_1` (
+    IN `inNormalizedEmail` VARCHAR(255)
+)
+BEGIN
+    DECLARE EXIT HANDLER FOR SQLEXCEPTION
+        BEGIN
+        ROLLBACK;
+        RESIGNAL;
+    END;
+
+    START TRANSACTION;
+
+    -- Don't delete any email that are considered primary emails
+	SET @isPrimary = 0;
+    SELECT COUNT(*) INTO @isPrimary FROM accounts WHERE normalizedEmail = inNormalizedEmail;
+    IF @isPrimary > 0 THEN
+      SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Can not delete a primary email address.';
+    END IF;
+
+	SELECT COUNT(*) INTO @isPrimary FROM emails WHERE normalizedEmail = inNormalizedEmail AND isPrimary = 1;
+    IF @isPrimary > 0 THEN
+      SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Can not delete a primary email address.';
+    END IF;
+
+	DELETE FROM emails WHERE normalizedEmail = inNormalizedEmail;
```

Comment:
If we assume that users will *usually* be doing the right thing, then you might be able to make this more efficient in the common case by using an "ask forgiveness, not permission" style:

* Do a `DELETE FROM emails WHERE normalizedEmail = inNormalizedEmail AND NOT isPrimary`
* Check how many rows were deleted.  If zero rows were deleted then:
  * Check whether it exists with `isPrimary=1`, error out if so
  * Check whether it exists in the main `accounts` table, error out if so
  * Otherwise fall through, reporting that zero rows were deleted

You could also move that "check what went wrong" logic out of the stored procedure and into the javascript method if it would simplify the code some.